tarpc/src/protocol/server.rs

```diff
@@ -212,29 +224,33 @@ pub trait Serve: Send + Sync + Sized {
     fn serve(&self, request: Self::Request) -> Self::Reply;
 
     /// spawn
-    fn spawn<A>(self, addr: A) -> io::Result<ServeHandle>
-        where A: ToSocketAddrs,
-              Self: 'static,
+    fn spawn<T>(self, transport: T) -> io::Result<ServeHandle<<T::Listener as Listener>::Dialer>>
+        where T: Transport,
+              Self: 'static
     {
-        self.spawn_with_config(addr, Config::default())
+        self.spawn_with_config(transport, Config::default())
     }
 
     /// spawn
-    fn spawn_with_config<A>(self, addr: A, config: Config) -> io::Result<ServeHandle>
-        where A: ToSocketAddrs,
-              Self: 'static,
+    fn spawn_with_config<T>(self,
+                            transport: T,
+                            config: Config)
+                            -> io::Result<ServeHandle<<T::Listener as Listener>::Dialer>>
+        where T: Transport,
+              Self: 'static
     {
-        let listener = try!(TcpListener::bind(&addr));
-        let addr = try!(listener.local_addr());
-        info!("spawn_with_config: spinning up server on {:?}", addr);
+        let listener = try!(transport.bind());
+        let dialer = try!(listener.dialer());
+        info!("spawn_with_config: spinning up server.");
```

Comment:
Let's just wait for specialization to land on stable, IMO.
On Mar 2, 2016 9:21 PM, "shaladdle" <notifications@github.com> wrote:

> In tarpc/src/protocol/server.rs
> <https://github.com/google/tarpc/pull/27#discussion_r54837191>:
>
> >      {
> > -        let listener = try!(TcpListener::bind(&addr));
> > -        let addr = try!(listener.local_addr());
> > -        info!("spawn_with_config: spinning up server on {:?}", addr);
> > +        let listener = try!(transport.bind());
> > +        let dialer = try!(listener.dialer());
> > +        info!("spawn_with_config: spinning up server.");
>
> Hmm, now that I see this, it kind of sucks to lose the ability to be able
> to say the address or 'location' of the server. Maybe Listener should
> impl Display or something? Or have an listen_addr method?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/google/tarpc/pull/27/files#r54837191>.
>